satellite/gc/service.go

```diff
@@ -0,0 +1,171 @@
+type Config struct {
+	Interval time.Duration `help:"the time between each send of garbage collection filters to storage nodes" releaseDefault:"168h" devDefault:"10m"`
+	Enabled  bool          `help:"set if garbage collection is enabled or not" releaseDefault:"true" devDefault:"true"`
+	// value for InitialPieces currently based on average pieces per node
+	InitialPieces     int     `help:"the initial number of pieces expected for a storage node to have, used for creating a filter" releaseDefault:"400000" devDefault:"10"`
+	FalsePositiveRate float64 `help:"the false positive rate used for creating a filter" releaseDefault:"0.1" devDefault:"0.1"`
+	ConcurrentSends   int     `help:"the number of nodes to concurrently send bloom filters to" releaseDefault:"1" devDefault:"1"`
+}
+
+// Service implements the garbage collection service
+type Service struct {
+	log          *zap.Logger
+	Loop         *sync2.Cycle
+	metainfoloop *metainfo.Loop
+	transport    transport.Client
+	overlay      overlay.DB
+	config       Config
+
+	lastPieceCounts map[storj.NodeID]int
+
+	pieceCountsMutex sync.Mutex
+	pieceCounts      map[storj.NodeID]int
+}
+
+// RetainInfo contains info needed for a storage node to retain important data and delete garbage data
+type RetainInfo struct {
+	Filter       *bloomfilter.Filter
+	CreationDate time.Time
+	count        int
+}
+
+// NewService creates a new instance of the gc service
+func NewService(log *zap.Logger, transport transport.Client, overlay overlay.DB, loop *metainfo.Loop, config Config) *Service {
+	// TODO retrieve piece counts from overlay (when there is a column for them)
+	// var lastPieceCounts atomic.Value
+	// lastPieceCounts.Store(map[storj.NodeID]int{})
+	lastPieceCounts := make(map[storj.NodeID]int)
+
+	return &Service{
+		log:              log,
+		Loop:             sync2.NewCycle(config.Interval),
+		metainfoloop:     loop,
+		transport:        transport,
+		overlay:          overlay,
+		lastPieceCounts:  lastPieceCounts,
+		config:           config,
+		pieceCountsMutex: sync.Mutex{},
+	}
+}
+
+// Run starts the gc loop service
+func (service *Service) Run(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	return service.Loop.Run(ctx, func(ctx context.Context) error {
+		obs := NewObserver(service.log.Named("gc observer"), service.lastPieceCounts, service.config)
+
+		err := service.metainfoloop.Join(ctx, obs)
+		if err != nil {
+			return Error.Wrap(err)
+		}
+
+		err = service.Send(ctx, obs)
+		if err != nil {
+			return Error.Wrap(err)
+		}
+
+		return nil
+	})
+}
+
+// Send sends the piece retain requests to all storage nodes
+func (service *Service) Send(ctx context.Context, obs *Observer) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	service.pieceCounts = make(map[storj.NodeID]int)
+
+	limiter := sync2.NewLimiter(service.config.ConcurrentSends)
+	for id, retainInfo := range obs.retainInfos {
+		func(id storj.NodeID, retainInfo *RetainInfo) {
```

Comment:
This is the first time I see such `func` statement without preceding it with `go`. What does it exactly do?

Reply:
it is an anonymous function that runs immediately in the same thread, rather than starting a separate thread as a goroutine. It allows us to pass in `id` and `retainInfo` as arguments into a goroutine without having to worry about the for loop iterating before the goroutine starts

Reply:
Also, this has been removed in favor of a simpler solution where we just copy `id` and `retainInfo` to new variables before starting `limiter.Go`